[Quant][fx] Remove WEIGHT_INDEX_DICT and BIAS_INDEX_DICT (Part 2)

jcaip opened the pull request:
Summary:
- Finishes the second part of https://github.com/pytorch/pytorch/pull/83263
- Removes WEIGHT_INDEX_DICT and BIAS_INDEX_DICT from utils.py
- Moves two funcitons, `node_arg_is_weight` and `node_arg_is_bias` into utils.py from prepare.py
convert.py and _equalize.py now use node_arg_is_weight instead of the dictionaries
- Adds in quantization support for `F.groupnorm`.

Add in missing BackendPatternConfigs for layernorm, instancenorm, and groupnorm

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

Reviewers:

Subscribers:

Tasks:

Tags:

ghstack-source-id: 2b157e0dc4f1553be1f4813b4693db952e6fc558
Pull Request resolved: https://github.com/pytorch/pytorch/pull/83848

Fixes #83093

jcaip commented on torch/ao/ns/fx/mappings.py:
```diff
@@ -195,6 +195,10 @@ def get_base_name_to_sets_of_related_ops() -> Dict[str, Set[NSNodeTargetType]]:
         set([
             F.hardswish,
         ]),
+        # F.group_norm
```
I added this to get a test to pass, but I don't know if I need to do anything else other than modify the dict

jerryzh168 replied:
not sure why we need this, cc @vkuzo to check

andrewor14 commented:
Can you add (part 2) to the title to differentiate this from the previous PR?

andrewor14 commented on torch/ao/quantization/fx/utils.py:
```diff
@@ -38,43 +39,40 @@
-    torch.nn.functional.instance_norm : [4],
-}
+def node_arg_is_weight(node: Node, arg: Any, backend_config: BackendConfig) -> bool:
+    "Returns if node arg is weight"
```
nit: use 3 double quotes here:
```
"""Returns if ... """
```

andrewor14 commented on torch/ao/quantization/fx/_equalize.py:
```diff
@@ -305,9 +307,10 @@ def get_op_node_and_weight_eq_obs(
-    for i, node_arg in enumerate(op_node.args):
-        if i in WEIGHT_INDEX_DICT[op_node.target]:  # type: ignore[index]
+    assert(op_node.op == 'call_function')
+    backend_config = get_native_backend_config()
```
I think we should add a TODO here to pass in the backend_config here instead

andrewor14 reviewed (commented):
Looks great! Just left some minor comments

jerryzh168 commented on torch/ao/quantization/backend_config/_common_operator_config_utils.py:
```diff
@@ -350,16 +350,35 @@ def _get_default_op_configs(dtype_configs: List[DTypeConfig]) -> List[BackendPat
+    )
+
+    configs.append(
+        BackendPatternConfig(torch.nn.functional.group_norm)
```
oh, I think we added quantization support for group_norm here, group_norm was not supported in the BackendConfig previously (since we removed L353 and L356 only, there were no entry for group_norm there)

vkuzo replied:
Quantized group_norm, instance_norm{n}d, layer_norm is supported in fbgemm but not qnnpack/xnnpack.  It would be great to move these config entries to the fbgemm section.

jcaip replied:
@vkuzo Should I pull out the `nn.Modules` too for instanceNorm/LayerNorm or just the functionals?

vkuzo replied:
Sorry, I made an outdated and incorrect statement.  What I should have said is that group_norm, instance_norm{n}d, layer_norm have an optimized quantized implementation on fbgemm, and have a naive implementation on xnnpack/qnnpack.  Please ignore my comment above.

jerryzh168 reviewed (approved):
looks great, thanks for removing the hardcode!

andrewor14 commented on torch/ao/quantization/backend_config/fbgemm.py:
```diff
@@ -71,6 +72,49 @@
+    )
+
+    configs.append(
+        BackendPatternConfig(torch.nn.functional.instance_norm)
```
Per @vkuzo's comment [here](https://github.com/pytorch/pytorch/pull/83853#discussion_r956261560), should we move these back to `_common_operator_config_utils.py`?

andrewor14 commented on torch/ao/quantization/backend_config/fbgemm.py:
```diff
@@ -71,6 +72,49 @@
+    )
+
+    configs.append(
+        BackendPatternConfig(torch.nn.functional.group_norm)
```
I guess this changes functionality, since `group_norm` was previously not defined in `BackendConfig`? Maybe we should mention this in the PR description to be clear.

andrewor14 commented on torch/ao/quantization/backend_config/fbgemm.py:
```diff
@@ -71,6 +72,49 @@
+        torch.nn.InstanceNorm1d,
+        torch.nn.InstanceNorm2d,
+        torch.nn.InstanceNorm3d,
+        torch.nn.LayerNorm,
```
If we're adding functional `group_norm` anyway, should we also add `torch.nn.GroupNorm` here? @jerryzh168

jerryzh168 replied:
yeah I think we can do that since we have quantized module for these things: https://github.com/pytorch/pytorch/blob/master/torch/ao/nn/quantized/modules/normalization.py, maybe better to do this in a separate PR.

jcaip replied:
I can make a PR to add it in later.

jcaip committed:
Remove WEIGHT_INDEX_DICT and BIAS_INDEX_DICT

Summary:
Removes WEIGHT_INDEX_DICT and BIAS_INDEX_DICT from utils.py
Moves two funcitons, node_arg_is_weight and node_arg_is_bias into utils.py from prepare.py
convert.py and _equalize.py now use node_arg_is_weight instead of the dictionaries

Add in missing PatternConfigs for layernorm, instancenorm, and groupnorm

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

Reviewers:

Subscribers:

Tasks:

Tags:

ghstack-source-id: 5a3719bda2e7160e9cc9bf1161ca9531c73adce6
Pull Request resolved: https://github.com/pytorch/pytorch/pull/83848

andrewor14 reviewed (approved):
Looks great. Thanks!

jcaip commented:
@pytorchbot merge
